src/server/webkit.ts

```diff
@@ -236,36 +236,149 @@ function getMacVersion(): string {
+    return value;
+  }
+}
+
 function wrapTransportWithWebSocket(transport: ConnectionTransport, port: number) {
   const server = new ws.Server({ port });
-  let socket: ws | undefined;
   const guid = uuidv4();
+  const idMixer = new SequenceNumberMixer<{id: number, socket: ws}>();
+  const pendingBrowserContextCreations = new Set<number>();
+  const pendingBrowserContextDeletions = new Map<number, string>();
+  const browserContextIds = new Map<string, ws>();
+  const pageProxyIds = new Map<string, ws>();
+  const sockets = new Set<ws>();
 
-  server.on('connection', (s, req) => {
-    if (req.url !== '/' + guid) {
-      s.close();
+  transport.onmessage = message => {
+    const parsedMessage = JSON.parse(message);
+    if ('id' in parsedMessage) {
+      if (parsedMessage.id === -9999)
+        return;
+      // Process command response.
+      const value = idMixer.take(parsedMessage.id);
+      if (!value)
+        return;
+      const { id, socket } = value;
+
+      if (pendingBrowserContextCreations.has(parsedMessage.id)) {
+        // Browser.createContext response -> establish context attribution.
+        browserContextIds.set(parsedMessage.result.browserContextId, socket);
+        pendingBrowserContextCreations.delete(parsedMessage.id);
+      }
+
+      const deletedContextId = pendingBrowserContextDeletions.get(parsedMessage.id);
+      if (deletedContextId) {
+        // Browser.createContext response -> remove context attribution.
+        browserContextIds.delete(deletedContextId);
+        pendingBrowserContextDeletions.delete(parsedMessage.id);
+      }
+
+      parsedMessage.id = id;
+      socket.send(JSON.stringify(parsedMessage));
       return;
     }
-    if (socket) {
-      s.close(undefined, 'Multiple connections are not supported');
+
+    // Process notification response.
+    const { method, params, pageProxyId } = parsedMessage;
+    if (pageProxyId) {
+      const socket = pageProxyIds.get(pageProxyId);
+      if (!socket || socket.readyState === ws.CLOSING) {
+        // Drop unattributed messages on the floor.
+        return;
+      }
+      socket.send(message);
       return;
     }
-    socket = s;
-    s.on('message', message => transport.send(Buffer.from(message).toString()));
-    transport.onmessage = message => {
-      // We are not notified when socket starts closing, and sending messages to a closing
-      // socket throws an error.
-      if (s.readyState !== ws.CLOSING)
-        s.send(message);
-    };
-    s.on('close', () => {
-      socket = undefined;
-      transport.onmessage = undefined;
+    if (method === 'Browser.pageProxyCreated') {
+      const socket = browserContextIds.get(params.pageProxyInfo.browserContextId);
+      if (!socket || socket.readyState === ws.CLOSING) {
+        // Drop unattributed messages on the floor.
+        return;
+      }
+      pageProxyIds.set(params.pageProxyInfo.pageProxyId, socket);
+      socket.send(message);
```

Comment:
Ca we restructure this in a way that doesn't depend on particular socket error? Something like this:
```js
try {
  socket.send(message);
  pageProxyIds.set(params.pageProxyInfo.pageProxyId, socket);
} catch (e) {
...
}
```